ide/posts/2020-06-50-ghcide-and-nixpkgs.md

```diff
@@ -0,0 +1,321 @@
+3. Compile one yourself using cabal.
+
+So I was developing an obelisk project and wanted to use ghcide. The first way
+was the recommend one in the `ghcide` README, but sadly that didn‘t work
+because (as I learned later from Elliot Cameron) the obelisk project uses a
+patched ghc. No luck here. We actually tried to pass the patched `ghc` into the
+`ghcide-nix` derivation, but `haskell.nix` refuses to work with a not
+`haskell.nix` ghc. So this was getting complicated quickly.
+2. Was also not an option, since `haskellPackages.ghcide` was marked broken in
+   nixpkgs and ignoring the flag just gave me tons of version mismatch errors.
+   So that left option three: Compiling with cabal myself. That worked-ish,
+   showing me that it‘s at least possible to get ghcide to run, but it is a
+   very unclean and brittle solution.
+
+After deciding that one and three can‘t help me I resolved to figure out how to
+fix the second approach.
+
+## The nixpkgs Haskell ecosytem and dependency resolution
+
+So I was wondering: Why can cabal get the dependencies right and nixpkgs can‘t?
+So for those of you who don‘t know: Let me digress on a short exploration of
+the nixpkgs Haskell ecosystem.
+
+### Haskell dependency resolution in general
+
+Dependency resolution problems have a long history in Haskell. But today there
+are two solutions that both work quite well in general.
+
+1. Specify upper and lower bounds for every dependency in your cabal file and
+   let cabal figure out a build plan. The times of cabal hell are over and this
+   works quite well. Notably this is the way ghcide is supposed to be compiled
+   in general.
+2. Pin a stack LTS release for your dependencies and pin the version for
+   packages not on stackage.
+
+Now solution two is in some sense less complex to use, because at compile time
+you don‘t need to construct a build plan. Of course, as I said, today cabal can
+do this for you very smoothly, which is why I personally prefer the first
+   approach.
+
+### Haskell in nixpkgs - pkgs.hackagePackages
+
+But how does nixpkgs do it? Well basically solution two. Everyday a cronjob
+pulls a list of **all packages from a pinned stack LTS release** and creates a
+derivation for every one of them. It also pulls **all other packages from
+hackage** and creates a derivation for the **latest released version** of them.
+(This happens on the haskell-updates branch of nixpkgs which get‘s normally
+merged into nixpkgs master i.e. unstable once per week.) So then, you ask, how
+does cabal2nix do dependency resolution? Well the short form is, it doesn‘t.
+What I mean by that is: It completely ignores any version bounds given in a
+cabal file or a pinned stack LTS release. It will just **take the one version
+of every dependency that is present in nixpkgs** by the method I told you
+above.
+
+When I first learned about this I thought this was ludicrous. How can someone
+device a not-solution like this? This is prone to fail. And indeed it does. A
```

Comment:
device a not-solution? I don't understand that phrase.

Reply:
Oh, actually I am rewriting much of the post right now and have stripped this part already.

I meant "to devise" …